lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -150,13 +157,28 @@ static bool validateCodingKeysEnum(DerivedConformance &derived,
   for (auto *varDecl : derived.Nominal->getStoredProperties()) {
     if (!varDecl->isUserAccessible())
       continue;
-
-    properties[getVarNameForCoding(varDecl)] = varDecl;
+    // When we have invalid redeclaration we may endup with duplicated
+    // members here.
+    auto varIdentifier = getVarNameForCoding(varDecl);
+    if (properties.count(varIdentifier) != 0)
+      continue;
+      
+    properties[varIdentifier] = varDecl;
   }
-
+  
+  llvm::SetVector<Identifier> validatedProperties;
   bool propertiesAreValid = true;
   for (auto elt : codingKeysDecl->getAllElements()) {
-    auto it = properties.find(elt->getBaseIdentifier());
+    auto eltIdentifier = elt->getBaseIdentifier();
+    
+    // We are skipping duplicated members on properties
+    // so let's avoid to validate the same declaration for
+    // a duplicated identifier.
+    if (validatedProperties.count(eltIdentifier) != 0)
```

Comment:
Why do we need additional filtering? `properties` is already mapping unique identifiers to properties.

Reply:
But here we are iterating on coding keys enum elements which can have synthesized or user-declared duplicated keys, we just keep track of what coding keys were already validated.

Reply:
Oh, I see now. I think we can simplify invalid case handling altogether by checking `elt->isInvalid()`, and setting `propertiesAreValid = false` to trigger the conformance error (property redeclarations do not affect validation of the coding keys enum, but an invalid `case` does). Please add a test case for this as well.

Reply:
I'm not sure about this, although you mention "property redeclarations do not affect validation of the coding keys enum, but an invalid case does" on a redeclaration one of the elements is invalid... so we can't diagnose only based in the elt being invalid(we can endup diagnose no conformance for redeclaration). 
Also, could you provide some example where a coding key would be invalid? I'm really not sure how this case will be...

Reply:
No, you're right, a redeclaration shouldn't affect the conformance either way. I really want us to avoid the new `SetVector`, though.

Reply:
Humm... we could remove duplicates from `allElements` before iterate but that will create another set as well ...

Reply:
What if we populate a `llvm::DenseSet` of coding keys separately to keep this tidy (we always validate them all, order is not crucial), and rather than erasing the properties from the map upon successful validation (which is a linear operation inside a loop) we can simply check if the property identifier maps to a coding key.

Reply:
> erasing the properties from the map upon successful validation (which is a linear operation inside a loop)

That's surprising, I was pretty sure that `erase` in a `DenseMap` was O(1) ... 

> we populate a llvm::DenseSet (we always validate them all, order is not crucial)

Humm... except the no erase part, I like the idea of the `DenseSet`, I'll try some things :)

Reply:
> That's surprising, I was pretty sure that `erase` in a `DenseMap` was O(1) ..

Ah, right, it seems to be so for `SmallDenseMap`.

Can you also add a test case for a duplicate conding key? We shouldn't be emitting a conformance error now that we skip them.